Disable WebGL in worker threads.

mattwoodrow opened the pull request:
#### b8652a99205d242d1d4f4247fa3f91def16c9a69
<pre>
Disable WebGL in worker threads.
<a href="https://bugs.webkit.org/show_bug.cgi?id=251341">https://bugs.webkit.org/show_bug.cgi?id=251341</a>
&lt;rdar://103571674&gt;

Reviewed by Brent Fulgham.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* LayoutTests/platform/ios-wk2/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/259711@main">https://commits.webkit.org/259711@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/41d266aa9d555feb680126a31842ddff833156e8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105567 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14651 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38458 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114813 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174958 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16071 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5874 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97856 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111322 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12226 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95205 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39712 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94090 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26842 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81416 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/95321 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7938 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28197 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/93467 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5717 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8331 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4785 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30390 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14061 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47741 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/102177 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6711 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9957 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25473 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -868,3 +868,12 @@
+    || ((PLATFORM(IOS) || PLATFORM(MACCATALYST)) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 170000) \
+    || (PLATFORM(WATCHOS) && __WATCH_OS_VERSION_MIN_REQUIRED >= 100000) \
+    || (PLATFORM(APPLETV) && __TV_OS_VERSION_MIN_REQUIRED >= 170000))
+#define ENABLE_WEBGL_IN_WORKERS 1
```
It's a little weird to create this compile time flag to comment out code a runtime check. Are you seeking to make it impossible to turn the flag on?

You might consider keeping the above change, discarding the change in OffscreenCanvas.cpp and modifying `UnifiedWebPreferences.yaml` as follows:

```
AllowWebGLInWorkers:
  type: bool
  status: stable
  humanReadableName: "Allow WebGL in Web Workers"
  condition: ENABLE(WEBGL)
  defaultValue:
    WebKitLegacy:
      "ENABLE(WEBGL_IN_WORKERS)": true
      default: false
    WebKit:
      "ENABLE(WEBGL_IN_WORKERS)": true
      default: false
    WebCore:
      "ENABLE(WEBGL_IN_WORKERS)": true
      default: false
```

brentfulgham replied:
If you don't want people turning this flag on outside an Engineering build, you could change the `status` to `unstable` instead.

mattwoodrow replied:
You're absolutely right, that's a much better way to implement this.

I don't think it matters if people can enable it outside of Engineering builds, we just want to set what the default is for now.

brentfulgham reviewed (approved):
Original change will work, but I think it might be simpler to just change the way the Preferences are set up.

mattwoodrow committed:
Disable WebGL in worker threads.
https://bugs.webkit.org/show_bug.cgi?id=251341
<rdar://103571674>

Reviewed by Brent Fulgham.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* LayoutTests/platform/ios-wk2/TestExpectations:

Canonical link: https://commits.webkit.org/259711@main
